### PR TITLE
Add tagname checker

### PR DESCRIPTION
**Pull request by ComeBurguburu**
Using class name as selector can provoque bug if the selector is not a `select` or if the select hasn't `option` child.

### CODE REVIEW

**Comment by ComeBurguburu** [comment]
clean pull request of #2227

File: coffee/lib/abstract-chosen.coffee [hunk 1]
```diff
@@ -280,8 +281,19 @@ class AbstractChosen
     if /Android/i.test(window.navigator.userAgent)
       return false if /Mobile/i.test(window.navigator.userAgent)
     return true
+    
+  @tagname_is_valid: : (@form_field) ->
+  if @form_field.tagName.toUpperCase() != 'SELECT'
+    console.error 'Chosen error: invalid tagName ' + @form_field + ' is not a select'
```
Comment (stof) [inline coffee/lib/abstract-chosen.coffee]: putting a DOM element in the message looks weird. How does the string representation looks like ?

File: coffee/lib/abstract-chosen.coffee [hunk 2]
```diff
@@ -280,8 +281,19 @@ class AbstractChosen
...
+    return false
+
+  if @form_field.getElementsByTagName('option').length == 0
+    console.error 'Chosen error: ' + @form_field + ' has no options\'s childs '
```
Comment (stof) [inline coffee/lib/abstract-chosen.coffee]: what if ``console.error`` does not exist (IE for instance) ?

File: coffee/lib/abstract-chosen.coffee [hunk 3]
```diff
@@ -280,8 +281,19 @@ class AbstractChosen
     if /Android/i.test(window.navigator.userAgent)
       return false if /Mobile/i.test(window.navigator.userAgent)
     return true
+    
+  @tagname_is_valid: : (@form_field) ->
+  if @form_field.tagName.toUpperCase() != 'SELECT'
+    console.error 'Chosen error: invalid tagName ' + @form_field + ' is not a select'
+    return false
+
+  if @form_field.getElementsByTagName('option').length == 0
```
Comment (stof) [inline coffee/lib/abstract-chosen.coffee]: use ``@form_field.options`` instead to get the options of a select

File: coffee/lib/abstract-chosen.coffee (hunk 1 above)
Comment (ComeBurguburu) [inline coffee/lib/abstract-chosen.coffee]: JQuery version  should be $(form_field).selector
don't know how to wrote it in coffee script

File: coffee/lib/abstract-chosen.coffee (hunk 2 above)
Comment (ComeBurguburu) [inline coffee/lib/abstract-chosen.coffee]: if (window.console) {

}

Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: You can use [CoffeeScript's existential operator](http://coffeescript.org/#operators-existential) here instead:

```coffeescript
console?.error "..."
```

File: coffee/lib/abstract-chosen.coffee [hunk 4]
```diff
@@ -280,8 +281,21 @@ class AbstractChosen
...
+    
+  @tagname_is_valid: : (@form_field) ->
+  if @form_field.tagName.toUpperCase() != 'SELECT'
+    if window.console
+      console.error 'Chosen error: invalid tagName ' + @form_field.selector + ' is not a select'
+    return false
+
+  if @form_field.options.length == 0
+      if  window.console
+       console.error 'Chosen error: ' + @form_field.selector + ' has no options\'s childs '
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: Grammar/formatting:  `' has no <option> children'`

File: coffee/lib/abstract-chosen.coffee [hunk 5]
```diff
@@ -280,8 +281,19 @@ class AbstractChosen
...
+
+  if @form_field.options.length == 0
+    console?.error 'Chosen error: ' + @form_field.selector + ' has no options\'s childs '
+    return false
+  return true
 
   @default_multiple_text: "Select Some Options"
   @default_single_text: "Select an Option"
   @default_no_result_text: "No results match"
+
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: Please cut this extra line add.

File: coffee/lib/abstract-chosen.coffee [hunk 6]
```diff
@@ -280,8 +281,18 @@ class AbstractChosen
...
+  if @form_field.options.length == 0
+    console?.error 'Chosen error: ' + @form_field.selector + ' has no options\'s childs '
+    return false
+  return true
 
   @default_multiple_text: "Select Some Options"
   @default_single_text: "Select an Option"
   @default_no_result_text: "No results match"
-
+
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: Heh, and now snip the extra indent addition :)

**Comment by ComeBurguburu** [comment]
extra lines removed

File: coffee/lib/abstract-chosen.coffee [hunk 7]
```diff
@@ -280,8 +281,17 @@ class AbstractChosen
...
       return false if /Mobile/i.test(window.navigator.userAgent)
     return true
+    
+  @tagname_is_valid: : (@form_field) ->
+  if @form_field.tagName.toUpperCase() != 'SELECT'
+    console?.error 'Chosen error: invalid tagName ' + @form_field.selector + ' is not a select'
+    return false
+
+  if @form_field.options.length == 0
+    console?.error 'Chosen error: ' + @form_field.selector + ' has no options\'s childs '
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: Grammar/formatting:  `' has no <option> children'`

File: coffee/lib/abstract-chosen.coffee [hunk 8]
```diff
@@ -280,8 +281,17 @@ class AbstractChosen
     if /Android/i.test(window.navigator.userAgent)
       return false if /Mobile/i.test(window.navigator.userAgent)
     return true
+    
+  @tagname_is_valid: : (@form_field) ->
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: What's up with this syntax here?  Seems like there's an extra `:` + space.

Additionally, this shouldn't even compile correctly -- did you actually compile the source to test this change in a real Chosen instance?

Comment (ComeBurguburu) [inline coffee/lib/abstract-chosen.coffee]: yeah : is a typo
I correct identation (the problems were wrong identations)
I successfully compiled it with npm coffee 1.9.0

File: coffee/lib/abstract-chosen.coffee [hunk 9]
```diff
@@ -281,7 +282,15 @@ class AbstractChosen
       return false if /Mobile/i.test(window.navigator.userAgent)
     return true
 
+  @tagname_is_valid: (@form_field) ->
+    if @form_field.tagName.toUpperCase() != 'SELECT'
+      console?.error 'Chosen error: invalid tagName ' + @form_field.selector + ' is not a select' 
+      return false
+    if @form_field.options.length == 0
+      console?.error 'Chosen error: ' + @form_field.selector + ' has no options\'s childs '
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: Grammar/formatting:  `' has no <option> children'`

File: coffee/lib/abstract-chosen.coffee [hunk 10]
```diff
@@ -281,7 +282,15 @@ class AbstractChosen
...
+      return false
+    return true
+
   @default_multiple_text: "Select Some Options"
   @default_single_text: "Select an Option"
   @default_no_result_text: "No results match"
-
```
Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: There should still be an empty new line at the end of the file.  But it should be empty, i.e. not including the blank two spaces that you had before.

Comment (ComeBurguburu) [inline coffee/lib/abstract-chosen.coffee]: added

Comment (tjschuck) [inline coffee/lib/abstract-chosen.coffee]: You added it back with the two blank spaces -- see [the diff](https://github.com/harvesthq/chosen/pull/2231/files).  There needs to be a new line, no spaces, which should mean nothing showing in the diff for the last line.

**Comment by pfiller** [comment]
Thanks @ComeBurguburu.

I'm not seeing a bug if Chosen is called on a select with no options. Chosen initializes properly and you can still add options dynamically without issue. If you're seeing differently, please share a [fiddle](http://jsfiddle.net) showing the problem.

I'm not convinced that Chosen should need to worry that you're applying it to a select element. If you're using specific enough selectors, this really shouldn't be a common problem and the breakage is going to be obvious.

For those reasons, I'm going to close this.

**Comment by ComeBurguburu** [comment]
The bug is here: 
http://jsfiddle.net/e6ys6grb/1/

I explain, in my application I duplicate a chosen-select when user click on "add" button.
I used a class to change the select style but also as jQuery selector

The bug is due to the use of the option inherit_select_classes

To correct the bug I replace
```````````````````````````````````````````````````js
$(".test").chosen({inherit_select_classes: true,width: '200px'});
```````````````````````````````````````````````````
by
`````````````````````````````````````````````````````````js
$("select.test").chosen({inherit_select_classes: true,width: '200px'});

**Comment by pfiller** [comment]
You could also just apply Chosen after you've appended the clone:

```JavaScript
$(document.body).append($(".test").clone(false))
$(".test").chosen({inherit_select_classes: true,width: '200px'})
```

**Pull request closed without merging**